Feature: G602 Slice Bound Checking

morgenm opened the pull request:
This PR addresses this issue: #954 

This adds a new rule which checks slice bounds access (reslicing and indexing) to ensure that slices are not accessed out of bounds. This only works for slice bounds defined by a `make([]..., len)` or `make([]..., len, capacity)` where len/capacity is defined as a literal integer. For example it works for:
```go
s := make([]float, 10, 20)
```
But not for:
```go
s := make([]float, 10*3)
```
Since we don't evaluate expressions. It also will calculate and validate slice capacities for new slices made by reslicing when possible, such as:
```go
s := make([]float, 10)
x := s[5:]
```
The capacity for the new slice, `x`, will be validated in any subsequent usage. In addition, slices passed to functions are validated when possible by storing function-call-specific capacities. For example:
```go
func main() {
	s := make([]int, 0, 30)
	doStuff(s)
	x := make([]int, 20)
	y := x[10:]
	doStuff(y)
	z := y[5:]
	doStuff(z)
}

func doStuff(x []int) {
	newSlice := x[:10]
	fmt.Println(newSlice)
	newSlice2 := x[:6]
	fmt.Println(newSlice2)
}`
```
will cause an error on creation of `newSlice` and `newSlice2`, since when `z` is passed to it, it is accessed out of bounds.

morgenm commented:
Note: This solution is complex since it stores the constant capacities of all the slices made with `make()`. I am thinking of adding in a simple check for non-constant capacities (or capacities evaluated from expressions such as `make([]int, 5*5)`) by making sure the user adds in a bounds check such as `if index < cap(slice)`. If it is better to simplify the whole check by only validating that the user has checked bounds, without storing the actual constant capacities in the map as I did in this PR, I can change the code.

Let me know your thoughts.

ccojocar commented on rules/slice_bounds.go:
```diff
@@ -0,0 +1,373 @@
+	funcCallArgs    map[string][]*int64                 // Caps to load once a func declaration is scanned
+	issue.MetaData
+}
+
```
Please could you add a comment to this publicly exposed method?

ccojocar commented on rules/slice_bounds.go:
```diff
@@ -0,0 +1,373 @@
+func (s *sliceOutOfBounds) ID() string {
+	return s.MetaData.ID
+}
+
```
ditto

ccojocar commented on rules/slice_bounds.go:
```diff
@@ -0,0 +1,373 @@
+	}
+}
+
+// Get all calls for the current function
```
The godoc format prefixes the comment with the method/function name.

ccojocar commented on rules/slice_bounds.go:
```diff
@@ -0,0 +1,373 @@
+				calls = append(calls, callExpr)
+			}
+		} else {
+			calls = append(calls, callExpr)
```
Why do you still append a nil value here?

morgenm replied:
I was storing functions that were never called by another function (main()) as "nil", then every time a new function declaration was scanned, the current function would become the new nil value to store info about the current call. I now have replaced this with creating dummy calls when loading the arguments of a function declaration, so no need for nil value and it will be less error prone.

ccojocar commented on rules/slice_bounds.go:
```diff
@@ -0,0 +1,373 @@
+			for _, cap := range caps {
+				if lowestCap == nil {
+					lowestCap = cap
+				} else if *lowestCap > *cap {
```
Is missing a nil check here? Can cap be nil?

morgenm replied:
Yes it can; I just wrote some tests and it caused runtime errors because of this. Good catch! I am fixing this now.

ccojocar commented on rules/slice_bounds.go:
```diff
@@ -0,0 +1,373 @@
+// Load caps that were saved for a call to this function
+func (s *sliceOutOfBounds) loadArgCaps(funcDecl *ast.FuncDecl) {
+	sliceMap := make(map[string]*int64)
+	funcName := funcDecl.Name.Name
```
Can the name be nil? e.g. for an anonymous function? I would check it here at least.

morgenm replied:
Anonymous functions would be type FuncLit, so it should never be nil to my understanding, as I noted in an above comment.

ccojocar commented on rules/slice_bounds.go:
```diff
@@ -0,0 +1,373 @@
+			continue
+		}
+
+		paramName := params[it].Names[0].Name
```
I would check `Name[0]` that is not nil before accessing the `Name` filed.

morgenm replied:
Okay, I added that check.

ccojocar commented on rules/slice_bounds.go:
```diff
@@ -0,0 +1,373 @@
+	s.sliceCaps[nil] = sliceMap
+}
+
+// Matches calls to make() and stores the capacity of the new slice in the map to compare against future slice usage
```
method name prefix in the comment

ccojocar commented on rules/slice_bounds.go:
```diff
@@ -0,0 +1,373 @@
+	return caps
+}
+
+// Matches slice assignments, calculates capacity of slice if possible to store it in map
```
function name prefix missing in the comment

ccojocar commented on rules/slice_bounds.go:
```diff
@@ -0,0 +1,373 @@
+func (s *sliceOutOfBounds) matchSliceAssignment(node *ast.SliceExpr, sliceName string, ctx *gosec.Context) (*issue.Issue, error) {
+	// First do the normal match that verifies the slice expr is not out of bounds
+	if i, err := s.matchSliceExpr(node, ctx); err != nil {
+		return i, err
```
I would wrap the err with some more context before returning.

morgenm replied:
Okay I wrapped it using `fmt.Errorf`.

ccojocar commented on rules/slice_bounds.go:
```diff
@@ -0,0 +1,373 @@
+	return nil, nil
+}
+
+// NewSliceBoundsCheck attempts to find any slices being accessed out of bounds
```
Could you please fix the function name in the comment?

ccojocar reviewed (commented):
Thank you for this great contribution. I left a few comments. It would be great if you could fix them before we merge this rule. Thanks again!

ccojocar commented:
> Note: This solution is complex since it stores the constant capacities of all the slices made with make(). I am thinking of adding in a simple check for non-constant capacities (or capacities evaluated from expressions such as make([]int, 5*5)) by making sure the user adds in a bounds check such as if index < cap(slice). If it is better to simplify the whole check by only validating that the user has checked bounds, without storing the actual constant capacities in the map as I did in this PR, I can change the code.

I think the current solution which checks the constant bounds looks fine to me. It is non intrusive comparing with a bound check which is more enforced on user. 

My impression is, that one will really check the bounds at runtime if she wants to be defensive but most of people won't do it. This will cause a lot of warnings which most likely will get ignored.

morgenm commented:
I believe I addressed all the reviews. 

> I think the current solution which checks the constant bounds looks fine to me. It is non intrusive comparing with a bound check which is more enforced on user.
> 
> My impression is, that one will really check the bounds at runtime if she wants to be defensive but most of people won't do it. This will cause a lot of warnings which most likely will get ignored.

Yes, thinking on it some more I agree with you.

razor-1 commented:
It would be nice if this didn't fire when it was protected by something like `if len(slice) >= x`

franchb commented:
Yes, this rule fires a lot of false positives on code protected with something like `if len(slice) >= x`
